lib/configuration.js

```diff
@@ -15,150 +15,335 @@
+ * @memberof Configuration
+ * @private
+ * @function _assimilateProjectId
+ * @param {Error|Undefined} err - If applicable, the error that occured during 
+ *  the transaction
+ * @param {Null|String} projectId - The project id if successfully 
+ *  retrieved, null otherwise.
+ * @param {Function} cb - The original user callback to invoke with the project
+ *  id or error encountered during id capture
+ * @returns {Undefined} - does not return anything
  */
-function gatherConfiguration(initConfiguration) {
-
-  var projectId = null;
-  var reportUncaughtExceptions = true;
-  var serviceContext = null;
-  var key = null;
-  var shouldReportErrorsToAPI = false;
-
-  if (env.NODE_ENV === 'production') {
-
-    shouldReportErrorsToAPI = true;
+Configuration.prototype._assimilateProjectId = function(err, projectId, cb) {
+  if (err) {
+    this._checkLocalProjectId(cb);
   } else {
-
-    console.log([
-      'Warning: the NODE_ENV environmental variable is not set to',
-      '\'production\'. To send errors to the Stackdriver Error Reporting API',
-      'please set the NODE_ENV to \'production\'.'
-    ].join(' '));
+    this._projectId = projectId;
+    cb(null, this._projectId);
   }
-
-  if (isPlainObject(initConfiguration)) {
-
-    if (has(initConfiguration, 'projectId')) {
-
-      projectId = initConfiguration.projectId;
-    }
-
-    if (has(initConfiguration, 'reportUncaughtExceptions')) {
-
-      reportUncaughtExceptions = !!initConfiguration.reportUncaughtExceptions;
-    }
-
-    if (has(initConfiguration, 'key')) {
-
-      key = initConfiguration.key;
+};
+/**
+ * The _checkLocalProjectId function is responsible for determing whether the
+ * _projectId property was set by the metadata service and whether or not the
+ * _projectId property should/can be set with a environmental or runtime
+ * configuration variable. If, upon execution of the _checkLocalProjectId
+ * function, the _projectId property has already been set to a string then it is
+ * assumed that this property has been set with the metadata services response.
+ * The metadata value for the project id always take precedence over any other
+ * locally configured project id value. Given that the metadata service did not
+ * set the project id this function will defer next to the value set in the
+ * environment named `GCLOUD_PROJECT` if it is set and of type string. If this
+ * environmental variable is not set the function will defer to the
+ * _givenConfiguration property if it is of type object and has a string
+ * property named projectId. If none of these conditions are met then the
+ * _projectId property will be left at its default value.
+ * @memberof Configuration
+ * @private
+ * @function _checkLocalProjectId
+ * @param {Function} cb - The original user callback to invoke with the project
+ *  id or error encountered during id capture
+ * @returns {Undefined} - does not return anything
+ */
+Configuration.prototype._checkLocalProjectId = function(cb) {
```

Comment:
@ofrobots because getProjectId is async and having this internal function respond to the callback makes it easier to separate out is specific logic versus the public getter function.